modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+// all inital cracks have to be defined at t_start = t_end = 0
+MeshCut3DUserObject::MeshCut3DUserObject(const InputParameters & parameters)
+  : GeometricCutUserObject(parameters), _mesh(_subproblem.mesh())
+{
+  if (isParamValid("n_step_growth"))
+    _n_step_growth = getParam<unsigned int>("n_step_growth");
+  else
+    _n_step_growth = 0;
+
+  if (_n_step_growth == 0)
+    _grow = 0;
+  else
+    _grow = 1;
+
+  if (_grow && !isParamValid("size_control"))
+    mooseError("Crack growth needs size control");
+
+  if (isParamValid("size_control"))
+    _size_control = getParam<Real>("size_control");
+
+  std::string xfem_cut_mesh_file = getParam<std::string>("mesh_file");
+  _cut_mesh = libmesh_make_unique<ReplicatedMesh>(_communicator);
+  _cut_mesh->read(xfem_cut_mesh_file);
+
+  // test element type
+  auto elem_it = _cut_mesh->elements_begin();
+  auto elem_end = _cut_mesh->elements_end();
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+    unsigned int n_node = cut_elem->n_nodes();
+    if (n_node != 3)
+      mooseError("The input cut mesh should include tri elements only!");
+  }
+
+  // this is a test of the growth methods
+  // will be moved to other places in a later version
+  if (_grow)
+  {
+    _stop = 0;
+
+    findBoundaryNodes();
+    findBoundaryEdges();
+    sortBoundaryNodes();
+    refineBoundary();
+
+    // visualize crack growth through an output file
+    // will be updated to other formats, e.g. *.e in a later version
+    std::ofstream myfile;
+    myfile.open("mesh_grow.out");
+
+    for (unsigned int i = 0; i < _n_step_growth; ++i)
+    {
+      findActiveBoundaryNodes();
+
+      if (_stop != 1)
+      {
+        findActiveBoundaryDirection();
+        growFront();
+        sortFrontNodes();
+
+        if (_inactive_boundary.size() != 0)
+          findFrontIntersection();
+
+        refineFront();
+        triangulation();
+        join();
+      }
+
+      myfile << _cut_mesh->n_nodes() << std::endl;
+      myfile << _cut_mesh->n_elem() << std::endl;
+      auto node_begin = _cut_mesh->nodes_begin();
+      auto node_end = _cut_mesh->nodes_end();
+      auto node_it = node_begin;
+      unsigned int n = 0;
+      for (; node_it != node_end; ++node_it)
+      {
+        Node * cutnode = *node_it;
+        Point & this_point = *cutnode;
+        myfile << this_point(0) << " " << this_point(1) << " " << this_point(2) << " " << std::endl;
+      }
+      auto elem_begin2 = _cut_mesh->elements_begin();
+      auto elem_end2 = _cut_mesh->elements_end();
+      auto elem_it2 = elem_begin2;
+      n = 0;
+      for (; elem_it2 != elem_end2; ++elem_it2)
+      {
+        Elem * cut_elem = *elem_it2;
+        std::vector<unsigned int> tri;
+        for (unsigned int i = 0; i < cut_elem->n_nodes(); ++i)
+        {
+          tri.push_back(cut_elem->node(i));
+        }
+        myfile << tri[0] << " " << tri[1] << " " << tri[2] << std::endl;
+      }
+    }
+    myfile.close();
+  }
+}
```

Comment:
You really shouldn't be doing all of this in the constructor of the object. Yes, it works but we try to minimize the work an object does in the constructor in MOOSE. Instead the initial work should be performed in `initialSetup`. If that's not possible here (because it doesn't inherit from one of the main MOOSE systems, fine. Just a thought on being more consistent with MOOSE).